core/services/llo/observation/data_source.go

```diff
@@ -85,139 +93,226 @@ func NewDataSource(lggr logger.Logger, registry Registry, t Telemeter) llo.DataS
+					promObservationErrorCount.WithLabelValues(streamIDStr).Inc()
 					mu.Lock()
 					errs = append(errs, ErrObservationFailed{inner: err, streamID: streamID, reason: "failed to observe stream"})
 					mu.Unlock()
 					return
 				}
 
 				// cache the observed value
-				d.toCache(opts.ConfigDigest(), streamID, val, opts.OutCtx().SeqNr)
-			}
+				d.toCache(streamID, val)
+			}(streamID)
+		}
 
-			mu.Lock()
-			defer mu.Unlock()
+		wg.Wait()
 
-			successfulStreamIDs = append(successfulStreamIDs, streamID)
-			if val != nil {
-				streamValues[streamID] = val
-			}
-		}(streamID)
-	}
+		// After all Observations have returned, nothing else will be sent to the
+		// telemetry channel, so it can safely be closed
+		if telemCh != nil {
+			close(telemCh)
+		}
+
+		// Only log on errors or if VerboseLogging is turned on
+		if len(errs) > 0 || opts.VerboseLogging() {
+			elapsed := time.Since(now)
 
-	// Wait for all Observations to complete
-	wg.Wait()
+			slices.Sort(successfulStreamIDs)
+			sort.Slice(errs, func(i, j int) bool { return errs[i].streamID < errs[j].streamID })
 
-	// Only log on errors or if VerboseLogging is turned on
-	if len(errs) > 0 || opts.VerboseLogging() {
-		elapsed := time.Since(now)
+			failedStreamIDs := make([]streams.StreamID, len(errs))
+			errStrs := make([]string, len(errs))
+			for i, e := range errs {
+				errStrs[i] = e.String()
+				failedStreamIDs[i] = e.streamID
+			}
 
-		slices.Sort(successfulStreamIDs)
-		sort.Slice(errs, func(i, j int) bool { return errs[i].streamID < errs[j].streamID })
+			lggr = logger.With(lggr, "elapsed", elapsed, "nSuccessfulStreams",
+				len(successfulStreamIDs), "nFailedStreams", len(failedStreamIDs), "errs", errStrs)
 
-		failedStreamIDs := make([]streams.StreamID, len(errs))
-		errStrs := make([]string, len(errs))
-		for i, e := range errs {
-			errStrs[i] = e.String()
-			failedStreamIDs[i] = e.streamID
+			if opts.VerboseLogging() {
+				lggr = logger.With(lggr, "streamValues", streamValues)
+			}
+
+			if len(errs) == 0 && opts.VerboseLogging() {
+				lggr.Infow("Observation succeeded for all streamsToObserve")
+			} else if len(errs) > 0 {
+				lggr.Warnw("Observation failed for streamsToObserve")
+			}
 		}
 
-		lggr = logger.With(lggr, "elapsed", elapsed, "nSuccessfulStreams",
-			len(successfulStreamIDs), "nFailedStreams", len(failedStreamIDs), "errs", errStrs)
+		// Notify the caller that we've completed our first round of observations.
+		if !d.observationLoopStarted {
+			d.observationLoopStarted = true
+			close(loopStartedCh)
+		}
 
-		if opts.VerboseLogging() {
-			lggr = logger.With(lggr, "streamValues", streamValues)
+		// Clean up anything that came up during the loop.
+		for _, fn := range loopCleanUpFns {
+			fn()
 		}
 
-		if len(errs) == 0 && opts.VerboseLogging() {
-			lggr.Infow("Observation succeeded for all streams")
-		} else if len(errs) > 0 {
-			lggr.Warnw("Observation failed for streams")
+		select {
+		case <-ctx.Done():
```

Comment:
this is guranteed to always be the case as the context is being cancelled in the loop above:
```
	// Clean up anything that came up during the loop.
		for _, fn := range loopCleanUpFns {
			fn()
		}
```

Reply:
this means we are terminating the loop and no continuously updating the cache. drop this select, the loopCleanUpFns, and just cancel() and sleep.

Reply:
We need some way to stop the goroutine.

Reply:
Synced offline, implemented `Closer` interface.